python/lsst/meas/astrom/matchOptimisticB.py

```diff
@@ -0,0 +1,241 @@
+        self.log.info("Purged %d unsuable sources, leaving %d clean sources" % \
+            (numSources - numCleanSources, numCleanSources))
+
+        # goodSourceCat: sources that are good and not saturated
+        goodSources = afwTable.SourceCatalog(sourceCat.table)
+        goodSources.extend(s for s in cleanSourceCat if isGoodSource(s))
+        numGoodSources = len(goodSources)
+        self.log.info("Purged %d saturated sources, leaving %d good sources" % \
+            (numCleanSources - numGoodSources, numGoodSources))
+
+        del sourceCat # avoid accidentally using sourceCat; use cleanSourceCat or goodSources from now on
+        
+        self.log.info("Matching to %d/%d good input sources" % (len(goodSources), len(cleanSourceCat)))
+
+        minMatchedPairs = min(self.config.minMatchedPairs,
+                            int(self.config.minFracMatchedPairs * min([len(refCat), len(goodSources)])))
+
+        # match clean (possibly saturated) sources and then purge the saturated sources from the match list
+        matchList0 = self.doMatch(
+            refCat = refCat,
+            sourceCat = cleanSourceCat,
+            wcs = wcs,
+            refFluxField = refFluxField,
+            numCleanSources = numCleanSources,
+            minMatchedPairs = minMatchedPairs,
+            verbose = debug.verbose,
+        )
+        if matchList0 is not None:
+            matchList0 = [m for m in matchList0 
+                           if (not m.second.getCentroidFlag() and 
+                               not m.second.get("base_PixelFlags_flag_saturated"))]
+        else:
+            matchList0 = []
+
+        # match good (unsaturated) sources
+        if len(refCat) > len(cleanSourceCat) - len(goodSources):
+            matchList1 = self.doMatch(
+                refCat = refCat,
+                sourceCat = goodSources,
+                wcs = wcs,
+                refFluxField = refFluxField,
+                numCleanSources = numCleanSources,
+                minMatchedPairs = minMatchedPairs,
+                verbose = debug.verbose,
+            )
+        else:
+            matchList1 = []
+        if matchList1 == None:
+            matchList1 = []
+
+        if len(matchList0) == 0 and len(matchList1) == 0:
+            raise RuntimeError("Unable to match sources")
+
+        # Adopt matchList with more matches
+        if len(matchList0) > len(matchList1):
+            matchList = matchList0
+        else:
+            matchList = matchList1
+
+        if self.log: self.log.info("Matched %d sources" % len(matchList))
+        if len(matchList) < minMatchedPairs:
+            self.log.warn("Number of matches is smaller than request")
+
+        return pipeBase.Struct(
+            matchList = matchList,
+        )
+
+    @pipeBase.timeMethod
+    def doMatch(self, refCat, sourceCat, wcs, refFluxField, numCleanSources, minMatchedPairs, verbose):
+        """!Match sources to position reference stars
```

Comment:
Name and description don't give any indication of what makes this different from `matchObjectsToSources`.  It seems like this might be a private method?